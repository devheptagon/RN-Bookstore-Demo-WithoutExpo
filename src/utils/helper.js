import React from "react";
import { FormattedCurrency } from "react-native-globalize";
import { createStackNavigator, createAppContainer } from "react-navigation";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import AppConfig from "../config/appConfig";

export const formatMoney = amount => {
  return <FormattedCurrency value={amount} />;
};

export const getRandomInt = (min, max) => {
  const rnd = Math.floor(Math.random() * Math.floor(max));
  return rnd < min ? min : rnd;
};

export const buildQueryUrl = (keyword, page) => {
  const { apiBaseUrl, pageSize } = AppConfig;
  return `${apiBaseUrl}?q=intitle:${keyword}&startIndex=${(page - 1) *
    pageSize}&maxResults=${pageSize}`;
};

export const buildDTOList = data => {
  let result = [];
  data.map(d => {
    const { volumeInfo, saleInfo, id } = d;
    const {
      imageLinks,
      title,
      subtitle,
      authors,
      publishedDate,
      averageRating,
      description,
      industryIdentifiers
    } = volumeInfo;
    const { listPrice } = saleInfo;
    const isbn =
      industryIdentifiers &&
      industryIdentifiers.length &&
      industryIdentifiers[0];
    result.push({
      id,
      volumeInfo: {
        imageLinks,
        title,
        subtitle,
        authors,
        publishedDate,
        averageRating
      },
      saleInfo: { listPrice },
      description,
      isbn
    });
  });
  return result;
};

export const flattenArray = style => {
  let result = {};
  style.forEach(element => {
    result = { ...result, ...element };
  });
  return result;
};

export const delay = duration =>
  new Promise(resolve => setTimeout(resolve, duration));

export const calculateBasketItemCount = basket => {
  return basket ? basket.reduce((acc, item) => acc + item.count, 0) : 0;
};
