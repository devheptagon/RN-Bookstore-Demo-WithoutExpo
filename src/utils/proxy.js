import api from "axios";

const Proxy = {
  sendRequest: function(httpVerb, url, data) {
    return api({
      url,
      headers: {
        "content-type": "application/json"
      },
      method: httpVerb,
      responseType: "json",
      data
    })
      .then(response => response.data)
      .catch(error => {
        console.error(error);
        alert(error);
      });
  },
  get: function(path) {
    return this.sendRequest("GET", path);
  },
  post: function(path, data) {
    return this.sendRequest("POST", path, data);
  },
  patch: function(path, data) {
    return this.sendRequest("PATCH", path, data);
  },
  delete: function(path, data) {
    return this.sendRequest("DELETE", path, data);
  }
};

export default Proxy;
