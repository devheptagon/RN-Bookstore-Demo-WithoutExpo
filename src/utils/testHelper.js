import React from 'react'
import { createStackNavigator, createAppContainer } from "react-navigation";
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'


export const wrapWithMockReduxAndNavigator = (Component) => {
  const initialState = { appReducer: {}}
  const mockStore = configureStore()

  const AppNavigator = createStackNavigator({
    Component
  })
  const ComponentWithNavigation = createAppContainer(AppNavigator)
  return <Provider store={mockStore(initialState)}>
      <ComponentWithNavigation />
    </Provider>   
}