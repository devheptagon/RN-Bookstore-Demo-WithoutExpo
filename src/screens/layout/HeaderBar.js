import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Icon from "react-native-vector-icons/AntDesign";
import { Text, StyleSheet } from "react-native";
import { SearchBar } from "react-native-elements";
import { setFilter } from "../../services/repos/productRepo";
import FlexView from "../components/FlexView";
import BasketStatusIcon from "../basket/components/BasketStatusIcon";
import BackButton from "../components/BackButton";
import {
  Pages,
  FlexAlign,
  FlexDirections,
  IconNames
} from "../../constants/Enums";
import { calculateBasketItemCount } from "../../utils/helper";
import Theme from "../../constants/Theme";
import Texts from "../../constants/Texts";

export class HeaderBar extends React.Component {
  state = {
    search: ""
  };

  updateSearch = search => {
    this.setState({ search });

    if (this.timer) clearTimeout(this.timer);
    this.timer = setTimeout(() => setFilter(search), 1000);
  };

  componentWillUnmount() {
    if (this.timer) clearTimeout(this.timer);
  }

  renderTitleContent = () => {
    const { pageTitle } = this.props;
    switch (pageTitle) {
      case "":
      case Pages.Home:
        return null;
      case Pages.ProductDetail:
      case Pages.Basket:
      case Pages.Address:
        return <BackButton navigator={this.props.mainNavigator} />;
      default:
        return <Text style={styles.TitleText}>{pageTitle}</Text>;
    }
  };

  render() {
    const { mainNavigator, basket, pageTitle } = this.props;
    const searchBarVisibility =
      pageTitle === Pages.Home || pageTitle === "" ? "flex" : "none";
    const titleVisibility =
      pageTitle === Pages.Home || pageTitle === "" ? "none" : "flex";
    return (
      <FlexView
        flex={0}
        flexDirection={FlexDirections.Row}
        alignItems={FlexAlign.Center}
        justifyContent={FlexAlign.SpaceBetween}
        style={styles.Wrapper}
      >
        <SearchBar
          placeholder={Texts.SearchText}
          onChangeText={this.updateSearch}
          value={this.state.search}
          containerStyle={[
            styles.SearchBarContainer,
            { display: searchBarVisibility }
          ]}
        />
        <FlexView
          alignItems={FlexAlign.FlexStart}
          style={[styles.TitleContainer, { display: titleVisibility }]}
        >
          {this.renderTitleContent()}
        </FlexView>
        <BasketStatusIcon
          itemCount={calculateBasketItemCount(basket)}
          onPress={() => mainNavigator.navigate(Pages.Basket)}
        />
        <Icon
          name={IconNames.HamburgerMenu}
          size={25}
          style={styles.HamburgerMenu}
          onPress={() => mainNavigator.toggleDrawer()}
        />
      </FlexView>
    );
  }
}

const styles = StyleSheet.create({
  Wrapper: {
    backgroundColor: Theme.HeaderBar.BackgroundColor,
    borderColor: Theme.HeaderBar.BorderColor,
    borderBottomWidth: Theme.General.Border.Width
  },
  HamburgerMenu: {
    color: Theme.HeaderBar.IconColor,
    marginLeft: 15,
    marginTop: 5,
    marginRight: 15
  },
  TitleContainer: {
    padding: 12,
    color: Theme.HeaderBar.TextColor
  },
  TitleText: {
    fontSize: Theme.General.Text.FontXBig,
    fontWeight: Theme.General.Text.FontBold,
    color: Theme.HeaderBar.TextColor
  },
  SearchBarContainer: {
    flex: 1,
    backgroundColor: Theme.HeaderBar.SearchBoxBackgroundColor,
    borderWidth: 0,
    shadowColor: Theme.HeaderBar.SearchBoxShadowColor,
    borderBottomColor: Theme.HeaderBar.SearchBoxBackgroundColor,
    borderTopColor: Theme.HeaderBar.SearchBoxBackgroundColor,
    paddingTop: 5,
    paddingBottom: 5,
    paddingLeft: 0
  }
});

HeaderBar.propTypes = {
  mainNavigator: PropTypes.object,
  basket: PropTypes.array,
  pageTitle: PropTypes.string
};

const mapStateToProps = state => {
  const {
    appReducer: { mainNavigator, pageTitle, basket }
  } = state;
  return {
    mainNavigator,
    pageTitle,
    basket
  };
};

export default connect(mapStateToProps)(HeaderBar);
