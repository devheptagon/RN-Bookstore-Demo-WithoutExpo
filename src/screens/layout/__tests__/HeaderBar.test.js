import React from 'react';
import { HeaderBar } from '../HeaderBar';
import renderer from 'react-test-renderer';
import { wrapWithMockReduxAndNavigator } from '../../../utils/testHelper'

test('HeaderBar renders correctly', () => {
  const component = wrapWithMockReduxAndNavigator(HeaderBar)
  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});