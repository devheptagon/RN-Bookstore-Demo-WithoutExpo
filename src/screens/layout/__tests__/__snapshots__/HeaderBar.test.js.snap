// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderBar renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    enabled={false}
    hitSlop={
      Object {
        "right": 50,
      }
    }
    maxDeltaY={20}
    minOffsetX={5}
    onGestureEvent={[Function]}
    onHandlerStateChange={[Function]}
  >
    <View
      style={
        Object {
          "backgroundColor": "#eee",
          "flex": 1,
          "flexDirection": "column-reverse",
          "overflow": "hidden",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          pointerEvents="auto"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "paddingTop": 64,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
                Object {
                  "translateY": 0,
                },
              ],
            }
          }
        >
          <View
            accessibilityElementsHidden={false}
            style={
              Object {
                "backgroundColor": "#fff",
                "flex": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#333",
                  "borderBottomWidth": 1,
                  "borderColor": "#ccc",
                  "flex": 0,
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                }
              }
            >
              <View
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "borderBottomColor": "transparent",
                    "borderBottomWidth": 1,
                    "borderTopColor": "transparent",
                    "borderTopWidth": 1,
                    "borderWidth": 0,
                    "display": "none",
                    "flex": 1,
                    "padding": 8,
                    "paddingBottom": 5,
                    "paddingLeft": 0,
                    "paddingTop": 5,
                    "shadowColor": "#fff",
                  }
                }
              >
                <View
                  style={
                    Object {
                      "paddingHorizontal": 0,
                      "width": "100%",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "#303337",
                        "borderBottomWidth": 0,
                        "borderColor": "#86939e",
                        "borderRadius": 3,
                        "flexDirection": "row",
                        "minHeight": 30,
                        "overflow": "hidden",
                        "transform": Array [
                          Object {
                            "translateX": 0,
                          },
                        ],
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "height": 40,
                          "justifyContent": "center",
                          "marginLeft": 8,
                        }
                      }
                    >
                      <View>
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "backgroundColor": "transparent",
                              "justifyContent": "center",
                            }
                          }
                          theme={
                            Object {
                              "colors": Object {
                                "disabled": "hsl(208, 8%, 90%)",
                                "divider": "#bcbbc1",
                                "error": "#ff190c",
                                "grey0": "#393e42",
                                "grey1": "#43484d",
                                "grey2": "#5e6977",
                                "grey3": "#86939e",
                                "grey4": "#bdc6cf",
                                "grey5": "#e1e8ee",
                                "greyOutline": "#bbb",
                                "platform": Object {
                                  "android": Object {
                                    "error": "#f44336",
                                    "primary": "#2196f3",
                                    "secondary": "#9C27B0",
                                    "success": "#4caf50",
                                    "warning": "#ffeb3b",
                                  },
                                  "ios": Object {
                                    "error": "#ff3b30",
                                    "primary": "#007aff",
                                    "secondary": "#5856d6",
                                    "success": "#4cd964",
                                    "warning": "#ffcc00",
                                  },
                                },
                                "primary": "#2089dc",
                                "searchBg": "#303337",
                                "secondary": "#8F0CE8",
                                "success": "#52c41a",
                                "warning": "#faad14",
                              },
                            }
                          }
                          underlayColor="white"
                        >
                          <Text />
                        </View>
                      </View>
                    </View>
                    <TextInput
                      allowFontScaling={true}
                      onBlur={[Function]}
                      onChangeText={[Function]}
                      onClear={[Function]}
                      onFocus={[Function]}
                      placeholder="Type Here..."
                      placeholderTextColor="#86939e"
                      platform="default"
                      style={
                        Object {
                          "alignSelf": "center",
                          "color": "#86939e",
                          "flex": 1,
                          "fontSize": 18,
                          "marginLeft": 10,
                          "minHeight": 40,
                        }
                      }
                      testID="searchInput"
                      underlineColorAndroid="transparent"
                      value=""
                    />
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "height": 40,
                          "justifyContent": "center",
                          "marginLeft": 15,
                          "marginRight": 8,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "flexDirection": "row",
                          }
                        }
                      />
                    </View>
                  </View>
                </View>
              </View>
              <View
                style={
                  Object {
                    "alignItems": "flex-start",
                    "color": "#fff",
                    "display": "flex",
                    "flex": 1,
                    "flexDirection": "column",
                    "padding": 12,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#fff",
                      "fontSize": 18,
                      "fontWeight": "bold",
                    }
                  }
                />
              </View>
              <View
                accessible={true}
                isTVSelectable={true}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={null}
              >
                <View
                  flexDirection="row"
                  style={Object {}}
                >
                  <Text />
                </View>
              </View>
              <Text />
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        pointerEvents="box-none"
        style={
          Object {
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#FFF",
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <View
            onLayout={[Function]}
            pointerEvents="box-none"
            style={
              Object {
                "backgroundColor": "#FFF",
                "borderBottomColor": "#A7A7AA",
                "borderBottomWidth": 0.5,
                "height": 64,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 20,
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "bottom": 0,
                    "flexDirection": "row",
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              >
                <View
                  pointerEvents="box-none"
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "bottom": 0,
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "left": 0,
                      "opacity": 1,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                >
                  <Text
                    accessibilityTraits="header"
                    allowFontScaling={false}
                    numberOfLines={1}
                    onLayout={[Function]}
                    style={
                      Object {
                        "color": "rgba(0, 0, 0, .9)",
                        "fontSize": 17,
                        "fontWeight": "600",
                        "marginHorizontal": 16,
                        "textAlign": "center",
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
